data_rentgen/server/utils/lineage_response.py

```diff
@@ -151,6 +169,46 @@ def _get_output_relations(outputs: dict[Any, OutputRow]) -> list[LineageOutputRe
     return sorted(relations, key=lambda x: (x.from_.kind, str(x.from_.id), str(x.to.id)))
 
 
+def _get_input_and_output_relations_with_dataset_granularity(
+    inputs: dict[Any, InputRow],
+    outputs: dict[Any, OutputRow],
+) -> tuple[list[LineageInputRelationV1], list[LineageOutputRelationV1]]:
+    i_relations = []
+    o_relations = []
+    for input_ in inputs.values():
+        for output in outputs.values():
+            if input_.run_id == output.run_id:
```

Comment:
This seems to be inefficient. Let's convert inputs & outputs to `dict[run_id, list[Input]]`, or even to `dict[tuple[run_id, dataset_id], Input]` all data summarized, and then iterate over matching keys.